jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,266 @@
+        )
+        self.solved_puzzle = jnp.arange(self.generator.grid_size**2).reshape(
+            (self.generator.grid_size, self.generator.grid_size)
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Resets the environment to an initial state."""
+        key, subkey = jax.random.split(key)
+        puzzle, empty_tile_position = self.generator(subkey)
+        state = State(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            key=key,
+        )
+        action_mask = self._get_valid_actions(empty_tile_position)
+        obs = Observation(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            action_mask=action_mask,
+        )
+        timestep = restart(observation=obs)
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """Updates the environment state after the agent takes an action."""
+        (updated_puzzle, updated_empty_tile_position) = self._move_empty_tile(
+            state.puzzle, state.empty_tile_position, action
+        )
+        # Check if the puzzle is solved
+        done = jnp.array_equal(updated_puzzle, self.solved_puzzle)
+
+        # Update the action mask
+        action_mask = self._get_valid_actions(updated_empty_tile_position)
+
+        next_state = State(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            key=state.key,
+        )
+        obs = Observation(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            action_mask=action_mask,
+        )
+
+        reward = self.reward_fn(state, action, next_state, self.solved_puzzle)
+
+        timestep = lax.cond(done, termination, transition, reward, obs)
+
+        return next_state, timestep
+
+    def _move_empty_tile(
+        self,
+        puzzle: chex.Array,
+        empty_tile_position: chex.Array,
+        action: chex.Array,
+    ) -> Tuple[chex.Array, chex.Array]:
+        """Moves the empty tile in the given direction and returns the updated puzzle and reward."""
+
+        # Compute the new position
+        new_empty_tile_position = empty_tile_position + self.movements[action]
+
+        # Predicate for the conditional
+        is_valid_move = jnp.all(
+            (new_empty_tile_position >= 0)
+            & (new_empty_tile_position < self.generator.grid_size)
+        )
+
+        def valid_move(puzzle: chex.Array) -> Tuple[chex.Array, Tuple[int, int]]:
+            # Swap the empty tile and the tile at the new position
+            updated_puzzle = puzzle.at[tuple(empty_tile_position)].set(
+                puzzle[tuple(new_empty_tile_position)]
+            )
+            updated_puzzle = updated_puzzle.at[tuple(new_empty_tile_position)].set(0)
```

Comment:
Can we make it clear that `0` is always the empty tile making it a constant? I rate make it a static variable in this class `EMPTY_TILE = 0`. Then you can use that here